### LSE Data Analytics Online Career Accelerator

# Course 2: Data Analytics using Python

## Assignment: Diagnostic Analysis using Python

You’ll be working with real-world data to address a problem faced by the National Health Service (NHS). The analysis will require you to utilise Python to explore the available data, create visualisations to identify trends, and extract meaningful insights to inform decision-making. 

### Preparing workstation

In [438]:
# Import the necessary libraries.
# using plotly express as the main library for visualisation
import pandas as pd
import numpy as np
import plotly.express as px

# Maximum column width to display
pd.options.display.max_colwidth = 200

# Optional - Ignore warnings.
import warnings
warnings.filterwarnings('ignore')

# Run all nodes interactively.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# 

# Assignment activity 1

### Insert proof of your GitHub repository. This can be a link or screenshot showing your repo.

## My GitHub repository.
## https://github.com/angjiashin/LSE_DA_NHS_analysis

![Repository](https://raw.githubusercontent.com/angjiashin/LSE_DA_NHS_analysis/main/images/github_repo.png)

# 

# Assignment activity 2

How many locations are there in the data set?

What are the five locations with the highest number of records?

How many service settings, context types, national categories, and appointment statuses are there?

In [439]:
# Import and sense-check the actual_duration.csv data set as ad.
ad = pd.read_csv('/Users/shin/Desktop/lse_daca/course_2/LSE_DA201_Assignment_files/actual_duration.csv')

# View the DataFrame.
ad.shape
ad.head(5)
ad.tail(5)

(137793, 8)

sub_icb_location_code sub_icb_location_ons_code  \
0                   00L                 E38000130   
1                   00L                 E38000130   
2                   00L                 E38000130   
3                   00L                 E38000130   
4                   00L                 E38000130   

                        sub_icb_location_name icb_ons_code region_ons_code  \
0  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
1  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
2  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
3  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
4  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   

  appointment_date         actual_duration  count_of_appointments  
0        01-Dec-21           31-60 Minutes                    364  
1        01-Dec-21           21-30 Minutes                    619  
2        01-Dec-21            6-10 Minutes                   1698  
3        01-Dec-21  Unknown / Data Quality                   1277  
4        01-Dec-21           16-20 Minutes                    730

sub_icb_location_code sub_icb_location_ons_code  \
137788                 X2C4Y                 E38000254   
137789                 X2C4Y                 E38000254   
137790                 X2C4Y                 E38000254   
137791                 X2C4Y                 E38000254   
137792                 X2C4Y                 E38000254   

                 sub_icb_location_name icb_ons_code region_ons_code  \
137788  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   
137789  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   
137790  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   
137791  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   
137792  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   

       appointment_date actual_duration  count_of_appointments  
137788        30-Jun-22   31-60 Minutes                    430  
137789        30-Jun-22   21-30 Minutes                    751  
137790        30-Jun-22   16-20 Minutes                    921  
137791        30-Jun-22   11-15 Minutes                   1439  
137792        30-Jun-22     1-5 Minutes                   1739

The actual_duration data set has 137,783 rows and 8 columns.

In [440]:
# Determine whether there are missing values.
ad.isnull().sum().sort_values(ascending = False)

sub_icb_location_code        0
sub_icb_location_ons_code    0
sub_icb_location_name        0
icb_ons_code                 0
region_ons_code              0
appointment_date             0
actual_duration              0
count_of_appointments        0
dtype: int64

Based on the shape and null value counts, it appears that there are no missing or null values.

In [441]:
# Determine the metadata of the data set.
ad.info()
print(ad.columns.tolist())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137793 entries, 0 to 137792
Data columns (total 8 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   sub_icb_location_code      137793 non-null  object
 1   sub_icb_location_ons_code  137793 non-null  object
 2   sub_icb_location_name      137793 non-null  object
 3   icb_ons_code               137793 non-null  object
 4   region_ons_code            137793 non-null  object
 5   appointment_date           137793 non-null  object
 6   actual_duration            137793 non-null  object
 7   count_of_appointments      137793 non-null  int64 
dtypes: int64(1), object(7)
memory usage: 8.4+ MB
['sub_icb_location_code', 'sub_icb_location_ons_code', 'sub_icb_location_name', 'icb_ons_code', 'region_ons_code', 'appointment_date', 'actual_duration', 'count_of_appointments']


The actual_duration data set consist of 2 data types: object 
('sub_icb_location_code', 'sub_icb_location_ons_code', 
'sub_icb_location_name', 'icb_ons_code', 'region_ons_code', 'appointment_date', 'actual_duration') 
and integer.('count_of_appointments')

In [442]:
# Determine the descriptive statistics of the data set.
ad.describe().T

count         mean          std  min    25%    50%  \
count_of_appointments  137793.0  1219.080011  1546.902956  1.0  194.0  696.0   

                          75%      max  
count_of_appointments  1621.0  15400.0

In [443]:
# Import and sense-check the appointments_regional.csv data set as ar.
ar = pd.read_csv('/Users/shin/Desktop/lse_daca/course_2/LSE_DA201_Assignment_files/appointments_regional.csv') 

# View the DataFrame.
ar.shape
ar.head(5)
ar.tail(5)

(596821, 7)

icb_ons_code appointment_month appointment_status hcp_type appointment_mode  \
0    E54000034           2020-01           Attended       GP     Face-to-Face   
1    E54000034           2020-01           Attended       GP     Face-to-Face   
2    E54000034           2020-01           Attended       GP     Face-to-Face   
3    E54000034           2020-01           Attended       GP     Face-to-Face   
4    E54000034           2020-01           Attended       GP     Face-to-Face   

  time_between_book_and_appointment  count_of_appointments  
0                             1 Day                   8107  
1                    15  to 21 Days                   6791  
2                       2 to 7 Days                  20686  
3                    22  to 28 Days                   4268  
4                     8  to 14 Days                  11971

icb_ons_code appointment_month appointment_status hcp_type  \
596816    E54000050           2022-06            Unknown  Unknown   
596817    E54000050           2022-06            Unknown  Unknown   
596818    E54000050           2022-06            Unknown  Unknown   
596819    E54000050           2022-06            Unknown  Unknown   
596820    E54000050           2022-06            Unknown  Unknown   

       appointment_mode time_between_book_and_appointment  \
596816          Unknown                       2 to 7 Days   
596817          Unknown                    22  to 28 Days   
596818          Unknown                     8  to 14 Days   
596819          Unknown                 More than 28 Days   
596820          Unknown                          Same Day   

        count_of_appointments  
596816                     21  
596817                      8  
596818                     28  
596819                     17  
596820                     10

The appointments_regional data set has 596,821 rows and 7 columns.

In [444]:
# Determine whether there are missing values.
ar.isnull().sum().sort_values(ascending = False)

icb_ons_code                         0
appointment_month                    0
appointment_status                   0
hcp_type                             0
appointment_mode                     0
time_between_book_and_appointment    0
count_of_appointments                0
dtype: int64

Based on the shape and null value counts, it appears that there are no missing or null values.

In [445]:
# Determine the metadata of the data set.
ar.info()
print(ar.columns.tolist())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596821 entries, 0 to 596820
Data columns (total 7 columns):
 #   Column                             Non-Null Count   Dtype 
---  ------                             --------------   ----- 
 0   icb_ons_code                       596821 non-null  object
 1   appointment_month                  596821 non-null  object
 2   appointment_status                 596821 non-null  object
 3   hcp_type                           596821 non-null  object
 4   appointment_mode                   596821 non-null  object
 5   time_between_book_and_appointment  596821 non-null  object
 6   count_of_appointments              596821 non-null  int64 
dtypes: int64(1), object(6)
memory usage: 31.9+ MB
['icb_ons_code', 'appointment_month', 'appointment_status', 'hcp_type', 'appointment_mode', 'time_between_book_and_appointment', 'count_of_appointments']


The appointments_regional data set consist of 2 data types: object 
('icb_ons_code', 'appointment_month', 
'appointment_status', 'hcp_type', 'appointment_mode', 'time_between_book_and_appointment') 
and integer.('count_of_appointments')

In [446]:
# Determine the descriptive statistics of the data set.
ar.describe().T

count         mean          std  min  25%   50%  \
count_of_appointments  596821.0  1244.601857  5856.887042  1.0  7.0  47.0   

                         75%       max  
count_of_appointments  308.0  211265.0

In [447]:
# Import and sense-check the national_categories.xlsx data set as nc.
nc = pd.read_excel('/Users/shin/Desktop/lse_daca/course_2/LSE_DA201_Assignment_files/national_categories.xlsx') 

# View the DataFrame.
nc.shape
nc.head(5)
nc.tail(5)

(817394, 8)

appointment_date icb_ons_code                       sub_icb_location_name  \
0       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
1       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
2       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
3       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
4       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   

        service_setting            context_type  \
0  Primary Care Network  Care Related Encounter   
1                 Other  Care Related Encounter   
2      General Practice  Care Related Encounter   
3      General Practice  Care Related Encounter   
4      General Practice  Care Related Encounter   

                        national_category  count_of_appointments  \
0  Patient contact during Care Home Round                      3   
1                         Planned Clinics                      7   
2                              Home Visit                     79   
3              General Consultation Acute                    725   
4            Structured Medication Review                      2   

  appointment_month  
0           2021-08  
1           2021-08  
2           2021-08  
3           2021-08  
4           2021-08

appointment_date icb_ons_code           sub_icb_location_name  \
817389       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   
817390       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   
817391       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   
817392       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   
817393       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   

                  service_setting            context_type  \
817389  Extended Access Provision  Care Related Encounter   
817390  Extended Access Provision  Care Related Encounter   
817391  Extended Access Provision  Care Related Encounter   
817392  Extended Access Provision  Care Related Encounter   
817393  Extended Access Provision  Care Related Encounter   

                   national_category  count_of_appointments appointment_month  
817389   Unplanned Clinical Activity                     12           2022-06  
817390               Planned Clinics                      4           2022-06  
817391    Planned Clinical Procedure                     92           2022-06  
817392  General Consultation Routine                      4           2022-06  
817393    General Consultation Acute                     19           2022-06

The national_categories data set has 817,394 rows and 8 columns.

In [448]:
# Determine whether there are missing values.
nc.isnull().sum().sort_values(ascending = False)

appointment_date         0
icb_ons_code             0
sub_icb_location_name    0
service_setting          0
context_type             0
national_category        0
count_of_appointments    0
appointment_month        0
dtype: int64

Based on the shape and null value counts, it appears that there are no missing or null values.

In [449]:
# Determine the metadata of the data set.
nc.info()
print(nc.columns.tolist())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 817394 entries, 0 to 817393
Data columns (total 8 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   appointment_date       817394 non-null  datetime64[ns]
 1   icb_ons_code           817394 non-null  object        
 2   sub_icb_location_name  817394 non-null  object        
 3   service_setting        817394 non-null  object        
 4   context_type           817394 non-null  object        
 5   national_category      817394 non-null  object        
 6   count_of_appointments  817394 non-null  int64         
 7   appointment_month      817394 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 49.9+ MB
['appointment_date', 'icb_ons_code', 'sub_icb_location_name', 'service_setting', 'context_type', 'national_category', 'count_of_appointments', 'appointment_month']


The national_categories data set consist of 3 data types:- 

1) object: 'appointment_date', 'icb_ons_code', 'sub_icb_location_name', 'service_setting', 'context_type', 'national_category', 'appointment_month'

2) integer: 'count_of_appointments'

3) datetime64[ns]: 'appointment_date'

In [450]:
# Determine the descriptive statistics of the data set.
nc.describe().T

count        mean        std  min  25%   50%    75%  \
count_of_appointments  817394.0  362.183684  1084.5766  1.0  7.0  25.0  128.0   

                           max  
count_of_appointments  16590.0

### Explore the data set

**Question 1:** How many locations are there in the data set?

In [451]:
# Determine the number of locations.
print(nc["sub_icb_location_name"].value_counts())
location_count = nc['sub_icb_location_name'].nunique()
print(f'There are {location_count} locations in the data set.')

NHS North West London ICB - W2U3Z              13007
NHS Kent and Medway ICB - 91Q                  12637
NHS Devon ICB - 15N                            12526
NHS Hampshire and Isle Of Wight ICB - D9Y0V    12171
NHS North East London ICB - A3A8R              11837
                                               ...  
NHS North East and North Cumbria ICB - 00N      4210
NHS Lancashire and South Cumbria ICB - 02G      4169
NHS Cheshire and Merseyside ICB - 01V           3496
NHS Cheshire and Merseyside ICB - 01T           3242
NHS Greater Manchester ICB - 00V                2170
Name: sub_icb_location_name, Length: 106, dtype: int64
There are 106 locations in the data set.


**Question 2:** What are the five locations with the highest number of records?



In [452]:
# Determine the top five locations based on record count.
print(f'The 5 locations with the highest number of records are: ')
print(nc['sub_icb_location_name'].value_counts().sort_values(ascending=False).head())

The 5 locations with the highest number of records are: 
NHS North West London ICB - W2U3Z              13007
NHS Kent and Medway ICB - 91Q                  12637
NHS Devon ICB - 15N                            12526
NHS Hampshire and Isle Of Wight ICB - D9Y0V    12171
NHS North East London ICB - A3A8R              11837
Name: sub_icb_location_name, dtype: int64


**Question 3:** How many service settings, context types, national categories, and appointment statuses are there?

In [453]:
# Determine the number of service settings.
print(nc["service_setting"].value_counts())
service_setting_count = nc["service_setting"].value_counts().count()
print(f' There are {service_setting_count} service settings.')

General Practice             359274
Primary Care Network         183790
Other                        138789
Extended Access Provision    108122
Unmapped                      27419
Name: service_setting, dtype: int64
 There are 5 service settings.


In [454]:
#  Percentage of General practice across all 5 service settings.
359274 /  (359274 + 183790 +  138789 + 108122 + 27419)

0.43953589089227474

In [455]:
# Determine the number of context types.
print(nc["context_type"].value_counts())
context_type_count = nc["context_type"].value_counts().count()
print(f' There are {context_type_count} context types.')

Care Related Encounter    700481
Inconsistent Mapping       89494
Unmapped                   27419
Name: context_type, dtype: int64
 There are 3 context types.


In [456]:
# Determine the number of national categories.
print(nc["national_category"].value_counts())
national_category_count = nc["national_category"].value_counts().count()
print(f' There are {national_category_count} national categories.')

Inconsistent Mapping                                                   89494
General Consultation Routine                                           89329
General Consultation Acute                                             84874
Planned Clinics                                                        76429
Clinical Triage                                                        74539
Planned Clinical Procedure                                             59631
Structured Medication Review                                           44467
Service provided by organisation external to the practice              43095
Home Visit                                                             41850
Unplanned Clinical Activity                                            40415
Patient contact during Care Home Round                                 28795
Unmapped                                                               27419
Care Home Visit                                                        26644

In [457]:
# Determine the number of appointment status.
print(ar["appointment_status"].value_counts())
appointment_status_count = ar["appointment_status"].value_counts().count()
print(f' There are {appointment_status_count} appointment statuses.')

Attended    232137
Unknown     201324
DNA         163360
Name: appointment_status, dtype: int64
 There are 3 appointment statuses.


# 

# Assignment activity 3

### Continue to explore the data and search for answers to more specific questions posed by the NHS.

**Question 1:** Between what dates were appointments scheduled? 

In [458]:
# View the first five rows of appointment_date for the ad DataFrame to determine the date format.
ad['appointment_date'].head()

0    01-Dec-21
1    01-Dec-21
2    01-Dec-21
3    01-Dec-21
4    01-Dec-21
Name: appointment_date, dtype: object

In [459]:
# View the first five rows of appointment_date for the nc DataFrame to determine the date format.
nc['appointment_date'].head()

0   2021-08-02
1   2021-08-02
2   2021-08-02
3   2021-08-02
4   2021-08-02
Name: appointment_date, dtype: datetime64[ns]

In [460]:
# Change the date format of ad['appointment_date'].
ad['appointment_date'] = pd.to_datetime(ad.appointment_date)

# View the DateFrame.
ad['appointment_date'].head()

0   2021-12-01
1   2021-12-01
2   2021-12-01
3   2021-12-01
4   2021-12-01
Name: appointment_date, dtype: datetime64[ns]

In [461]:
# Change the date format of nc['appointment_date'].
nc['appointment_date'] = pd.to_datetime(nc.appointment_date)

# View the DateFrame.
nc['appointment_date'].head()

0   2021-08-02
1   2021-08-02
2   2021-08-02
3   2021-08-02
4   2021-08-02
Name: appointment_date, dtype: datetime64[ns]

In [462]:
# Determine the minimum and maximum dates in the ad DataFrame.
# Use appropriate docstrings.
min_date = ad['appointment_date'].min()
max_date = ad['appointment_date'].max()
print(f'First date of scheduled appointments: {min_date}')
print(f'Last date of scheduled appointments: {max_date}')      

First date of scheduled appointments: 2021-12-01 00:00:00
Last date of scheduled appointments: 2022-06-30 00:00:00


In [463]:
# Determine the minimum and maximum dates in the nc DataFrame.
# Use appropriate docstrings.
min_date = nc['appointment_date'].min()
max_date = nc['appointment_date'].max()
print(f'First date of scheduled appointments: {min_date}')
print(f'Last date of scheduled appointments: {max_date}')      

First date of scheduled appointments: 2021-08-01 00:00:00
Last date of scheduled appointments: 2022-06-30 00:00:00


**Question 2:** Which service setting was the most popular for NHS North West London from 1 January to 1 June 2022?

In [464]:
# For each of these service settings, determine the number of records available for the period and the location. 
nc_subset = nc.loc[(nc['sub_icb_location_name'] == 'NHS North West London ICB - W2U3Z') & 
                   (nc['appointment_date'] >= '2022-01-01') & 
                   (nc['appointment_date'] <= '2022-06-01')]

# View the output.
print(nc_subset["service_setting"].value_counts().sort_values(ascending = False).head(1))

General Practice    2104
Name: service_setting, dtype: int64


**Question 3:** Which month had the highest number of appointments?

In [465]:
nc.head()

appointment_date icb_ons_code                       sub_icb_location_name  \
0       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
1       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
2       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
3       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
4       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   

        service_setting            context_type  \
0  Primary Care Network  Care Related Encounter   
1                 Other  Care Related Encounter   
2      General Practice  Care Related Encounter   
3      General Practice  Care Related Encounter   
4      General Practice  Care Related Encounter   

                        national_category  count_of_appointments  \
0  Patient contact during Care Home Round                      3   
1                         Planned Clinics                      7   
2                              Home Visit                     79   
3              General Consultation Acute                    725   
4            Structured Medication Review                      2   

  appointment_month  
0           2021-08  
1           2021-08  
2           2021-08  
3           2021-08  
4           2021-08

In [466]:
# Number of appointments per month == sum of count_of_appointments by month.
# Use the groupby() and sort_values() functions.
nc.groupby([
    nc['appointment_date'].dt.year.rename('year'), 
    nc['appointment_date'].dt.month.rename('month')])[['count_of_appointments']].sum().sort_values(by=
                                                                                                   'count_of_appointments',
                                                                                                    ascending = False).reset_index()

year  month  count_of_appointments
0   2021     11               30405070
1   2021     10               30303834
2   2022      3               29595038
3   2021      9               28522501
4   2022      5               27495508
5   2022      6               25828078
6   2022      1               25635474
7   2022      2               25355260
8   2021     12               25140776
9   2022      4               23913060
10  2021      8               23852171

From the dataframe above, November 2021 had the highest number of appointments.

**Question 4:** What was the total number of records per month?

In [467]:
# Total number of records per month.
nc.groupby('appointment_month')[['count_of_appointments']].sum().reset_index()

appointment_month  count_of_appointments
0            2021-08               23852171
1            2021-09               28522501
2            2021-10               30303834
3            2021-11               30405070
4            2021-12               25140776
5            2022-01               25635474
6            2022-02               25355260
7            2022-03               29595038
8            2022-04               23913060
9            2022-05               27495508
10           2022-06               25828078

# 

# Assignment activity 4

### Create visualisations and identify possible monthly and seasonal trends in the data.

### Objective 1
Create three visualisations indicating the number of appointments per month for service settings, context types, and national categories.

In [468]:
# Change the data type of the appointment month to string to allow for easier plotting.
nc['appointment_month'].astype(str)

0         2021-08
1         2021-08
2         2021-08
3         2021-08
4         2021-08
           ...   
817389    2022-06
817390    2022-06
817391    2022-06
817392    2022-06
817393    2022-06
Name: appointment_month, Length: 817394, dtype: object

In [469]:
# Aggregate on monthly level and determine the sum of records per month.
# create a seperate data set for each column.
# (service settings, context types, and national categories)
nc_ss = nc.groupby(['appointment_month', 'service_setting'])[['count_of_appointments']].sum().reset_index()
nc_ct = nc.groupby(['appointment_month', 'context_type'])[['count_of_appointments']].sum().reset_index()
nc_nc = nc.groupby(['appointment_month', 'national_category'])[['count_of_appointments']].sum().reset_index()

# View output.
nc_ss.head()
nc_ct.head()
nc_nc.head()

appointment_month            service_setting  count_of_appointments
0           2021-08  Extended Access Provision                 160927
1           2021-08           General Practice               21575852
2           2021-08                      Other                 449101
3           2021-08       Primary Care Network                 432448
4           2021-08                   Unmapped                1233843

appointment_month            context_type  count_of_appointments
0           2021-08  Care Related Encounter               20255235
1           2021-08    Inconsistent Mapping                2363093
2           2021-08                Unmapped                1233843
3           2021-09  Care Related Encounter               24404251
4           2021-09    Inconsistent Mapping                2782135

appointment_month  \
0           2021-08   
1           2021-08   
2           2021-08   
3           2021-08   
4           2021-08   

                                                     national_category  \
0  Care Home Needs Assessment & Personalised Care and Support Planning   
1                                                      Care Home Visit   
2                                                      Clinical Triage   
3                                           General Consultation Acute   
4                                         General Consultation Routine   

   count_of_appointments  
0                  29676  
1                  47583  
2                3704207  
3                4280920  
4                7756045

**Service settings:**

In [470]:
# Plot the appointments over the available date range, and review the service settings for months.
# Create a lineplot.
fig = px.line(nc_ss,
              x = 'appointment_month',
              y = 'count_of_appointments',
              color = 'service_setting',
              title = 'Total appointment across months and service settings', 
              labels = {'appointment_month': 'Month',
                        'count_of_appointments': 'Number of appointments',
                        'service_setting': 'Service Setting'})

fig.show()

**Context types:**

In [471]:
# Plot the appointments over the available date range, and review the context types for months.
# Create a lineplot.
fig = px.line(nc_ct,
              x = 'appointment_month',
              y = 'count_of_appointments',
              color = 'context_type',
              title = 'Total appointment across months and context types', 
              labels = {'appointment_month': 'Month',
                        'count_of_appointments': 'Number of appointments',
                        'context_type': 'Context Type'})

fig.show()

**National categories:**

In [472]:
# Plot the appointments over the available date range, and review the national categories for months.
# Create a lineplot.
fig = px.line(nc_nc,
              x = 'appointment_month',
              y = 'count_of_appointments',
              color = 'national_category',
              title = 'Total appointment across months and national categories', 
              labels = {'appointment_month': 'Month', 
                        'count_of_appointments': 'Number of appointments', 
                        'national_category': 'National Category'})
fig.show()

### Objective 2
Create four visualisations indicating the number of appointments for service setting per season. The seasons are summer (August 2021), autumn (October 2021), winter (January 2022), and spring (April 2022).

**Summer (August 2021):**

In [473]:
nc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 817394 entries, 0 to 817393
Data columns (total 8 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   appointment_date       817394 non-null  datetime64[ns]
 1   icb_ons_code           817394 non-null  object        
 2   sub_icb_location_name  817394 non-null  object        
 3   service_setting        817394 non-null  object        
 4   context_type           817394 non-null  object        
 5   national_category      817394 non-null  object        
 6   count_of_appointments  817394 non-null  int64         
 7   appointment_month      817394 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 49.9+ MB


In [474]:
# Create a separate data set that can be used in future weeks. 

nc_ss_day = nc.groupby(['appointment_date',
                        'appointment_month',
                        'service_setting'])[['count_of_appointments']].sum().reset_index()

nc_ss_summer = nc_ss_day.loc[nc_ss_day['appointment_month'] == '2021-08']

nc_ss_autumn = nc_ss_day.loc[nc_ss_day['appointment_month'] == '2021-10']

nc_ss_winter = nc_ss_day.loc[nc_ss_day['appointment_month'] == '2022-01']

nc_ss_spring = nc_ss_day.loc[nc_ss_day['appointment_month'] == '2022-04']

# View output.
nc_ss_summer
nc_ss_autumn
nc_ss_winter
nc_ss_spring

appointment_date appointment_month            service_setting  \
0         2021-08-01           2021-08  Extended Access Provision   
1         2021-08-01           2021-08           General Practice   
2         2021-08-01           2021-08                      Other   
3         2021-08-01           2021-08       Primary Care Network   
4         2021-08-01           2021-08                   Unmapped   
..               ...               ...                        ...   
150       2021-08-31           2021-08  Extended Access Provision   
151       2021-08-31           2021-08           General Practice   
152       2021-08-31           2021-08                      Other   
153       2021-08-31           2021-08       Primary Care Network   
154       2021-08-31           2021-08                   Unmapped   

     count_of_appointments  
0                      438  
1                     3411  
2                      401  
3                      323  
4                     1054  
..                     ...  
150                   8281  
151                1125584  
152                  23181  
153                  21789  
154                  62380  

[155 rows x 4 columns]

appointment_date appointment_month            service_setting  \
305       2021-10-01           2021-10  Extended Access Provision   
306       2021-10-01           2021-10           General Practice   
307       2021-10-01           2021-10                      Other   
308       2021-10-01           2021-10       Primary Care Network   
309       2021-10-01           2021-10                   Unmapped   
..               ...               ...                        ...   
455       2021-10-31           2021-10  Extended Access Provision   
456       2021-10-31           2021-10           General Practice   
457       2021-10-31           2021-10                      Other   
458       2021-10-31           2021-10       Primary Care Network   
459       2021-10-31           2021-10                   Unmapped   

     count_of_appointments  
305                   7107  
306                1130646  
307                  22639  
308                  22974  
309                  56201  
..                     ...  
455                   1113  
456                  17449  
457                   1147  
458                   1200  
459                   2863  

[155 rows x 4 columns]

appointment_date appointment_month            service_setting  \
765       2022-01-01           2022-01  Extended Access Provision   
766       2022-01-01           2022-01           General Practice   
767       2022-01-01           2022-01                      Other   
768       2022-01-01           2022-01       Primary Care Network   
769       2022-01-01           2022-01                   Unmapped   
..               ...               ...                        ...   
915       2022-01-31           2022-01  Extended Access Provision   
916       2022-01-31           2022-01           General Practice   
917       2022-01-31           2022-01                      Other   
918       2022-01-31           2022-01       Primary Care Network   
919       2022-01-31           2022-01                   Unmapped   

     count_of_appointments  
765                    246  
766                   3578  
767                    280  
768                    215  
769                    572  
..                     ...  
915                   9962  
916                1305082  
917                  25034  
918                  30509  
919                  43240  

[155 rows x 4 columns]

appointment_date appointment_month            service_setting  \
1215       2022-04-01           2022-04  Extended Access Provision   
1216       2022-04-01           2022-04           General Practice   
1217       2022-04-01           2022-04                      Other   
1218       2022-04-01           2022-04       Primary Care Network   
1219       2022-04-01           2022-04                   Unmapped   
...               ...               ...                        ...   
1360       2022-04-30           2022-04  Extended Access Provision   
1361       2022-04-30           2022-04           General Practice   
1362       2022-04-30           2022-04                      Other   
1363       2022-04-30           2022-04       Primary Care Network   
1364       2022-04-30           2022-04                   Unmapped   

      count_of_appointments  
1215                   8011  
1216                1085037  
1217                  21077  
1218                  26493  
1219                  36836  
...                     ...  
1360                   3416  
1361                  23675  
1362                   1148  
1363                   3150  
1364                   1844  

[150 rows x 4 columns]

In [475]:
# Look at August 2021 in more detail to allow a closer look.
# Create a lineplot.
fig = px.line(nc_ss_summer,
              x = 'appointment_date',
              y = 'count_of_appointments',
              color = 'service_setting',
              title = 'Total appointment across summer and service settings', 
              labels = {'appointment_date': 'day',
                        'count_of_appointments': 'Number of appointments',
                        'service_setting': 'Service Setting'})

fig.show()

**Autumn (October 2021):**

In [476]:
# Look at October 2021 in more detail to allow a closer look.
# Create a lineplot.
fig = px.line(nc_ss_autumn,
              x = 'appointment_date',
              y = 'count_of_appointments',
              color = 'service_setting',
              title = 'Total appointment across autumn and service settings', 
              labels = {'appointment_date': 'day',
                        'count_of_appointments': 'Number of appointments',
                        'service_setting': 'Service Setting'})

fig.show()

**Winter (January 2022):**

In [477]:
# Look at January 2022 in more detail to allow a closer look.
# Create a lineplot.
fig = px.line(nc_ss_winter,
              x = 'appointment_date',
              y = 'count_of_appointments',
              color = 'service_setting',
              title = 'Total appointment across winter and service settings', 
              labels = {'appointment_date': 'day',
                        'count_of_appointments': 'Number of appointments',
                        'service_setting': 'Service Setting'})
fig.show()

**Spring (April 2022):**

In [478]:
# Look at April 2022 in more detail to allow a closer look.
# Create a lineplot.
fig = px.line(nc_ss_spring,
              x = 'appointment_date',
              y = 'count_of_appointments',
              color = 'service_setting',
              title = 'Total appointment across spring and service settings', 
              labels = {'appointment_date': 'day',
                        'count_of_appointments': 'Number of appointments',
                        'service_setting': 'Service Setting'})
fig.show()

# 

# Assignment activity 5

### Analyse tweets from Twitter with hashtags related to healthcare in the UK.

In [479]:
# Load the data set.
tweets = pd.read_csv('/Users/shin/Desktop/lse_daca/course_2/LSE_DA201_Assignment_files/tweets.csv')

# View the DataFrame.
tweets.head()

tweet_id  \
0  1567629223795527681   
1  1567582846612553728   
2  1567582787070304256   
3  1567582767625428992   
4  1567582720460570625   

                                                                                                                                                                                           tweet_full_text  \
0  As Arkansas’ first Comprehensive Stroke Certified Center, UAMS provides Arkansans with access to the most advanced stoke care. Join us in our mission to make a difference in the health and well-be...   
1                                                              RT @AndreaGrammer: Work-life balance is at the foundation of how decisions are made and where #PremiseHealth is headed.  We're #hiring for…   
2                                                           RT @OntarioGreens: $10 billion can go a long way to fixing our broken #Healthcare system.\n\nYet Doug Ford would rather spend it ALL on a hig…   
3                                                             RT @modrnhealthcr: 🚨#NEW:🚨 Insurance companies are figuring out the best ways to collect information about members’ race and ethnicity data…   
4  ICYMI: Our recent blogs on Cybersecurity in Accounting https://t.co/4nnK0FiVVL and Digital Transformation in Healthcare Finance https://t.co/jIqn52lHD3 are a great read, take a look!\n\n#blogs #di...   

                                                                                                                                                                                            tweet_entities  \
0  {'hashtags': [{'text': 'Healthcare', 'indices': [253, 264]}], 'symbols': [], 'user_mentions': [], 'urls': [{'url': 'https://t.co/yw0cstfmSI', 'expanded_url': 'https://bit.ly/3BiSKbs', 'display_url...   
1  {'hashtags': [{'text': 'PremiseHealth', 'indices': [94, 108]}, {'text': 'hiring', 'indices': [127, 134]}], 'symbols': [], 'user_mentions': [{'screen_name': 'AndreaGrammer', 'name': 'Andrea Grammer...   
2  {'hashtags': [{'text': 'Healthcare', 'indices': [70, 81]}], 'symbols': [], 'user_mentions': [{'screen_name': 'OntarioGreens', 'name': 'Green Party of Ontario', 'id': 37115912, 'id_str': '37115912'...   
3  {'hashtags': [{'text': 'NEW', 'indices': [20, 24]}], 'symbols': [], 'user_mentions': [{'screen_name': 'modrnhealthcr', 'name': 'Modern Healthcare', 'id': 18935711, 'id_str': '18935711', 'indices':...   
4  {'hashtags': [{'text': 'blogs', 'indices': [184, 190]}, {'text': 'digitaltransformation', 'indices': [191, 213]}, {'text': 'cybersecurity', 'indices': [214, 228]}, {'text': 'accounting', 'indices'...   

                                                              tweet_entities_hashtags  \
0                                                                         #Healthcare   
1                                                             #PremiseHealth, #hiring   
2                                                                         #Healthcare   
3                                                                                #NEW   
4  #blogs, #digitaltransformation, #cybersecurity, #accounting, #finance, #healthcare   

                                         tweet_metadata  tweet_retweet_count  \
0  {'iso_language_code': 'en', 'result_type': 'recent'}                    0   
1  {'iso_language_code': 'en', 'result_type': 'recent'}                    2   
2  {'iso_language_code': 'en', 'result_type': 'recent'}                   39   
3  {'iso_language_code': 'en', 'result_type': 'recent'}                    5   
4  {'iso_language_code': 'en', 'result_type': 'recent'}                    0   

   tweet_favorite_count  tweet_favorited  tweet_retweeted tweet_lang  
0                     0            False            False         en  
1                     0            False            False         en  
2                     0            False            False         en  
3                     0            False            False         en  
4           

In [480]:
# Explore the metadata.
tweets.info()
print(tweets.columns.tolist())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1174 entries, 0 to 1173
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   tweet_id                 1174 non-null   int64 
 1   tweet_full_text          1174 non-null   object
 2   tweet_entities           1174 non-null   object
 3   tweet_entities_hashtags  1007 non-null   object
 4   tweet_metadata           1174 non-null   object
 5   tweet_retweet_count      1174 non-null   int64 
 6   tweet_favorite_count     1174 non-null   int64 
 7   tweet_favorited          1174 non-null   bool  
 8   tweet_retweeted          1174 non-null   bool  
 9   tweet_lang               1174 non-null   object
dtypes: bool(2), int64(3), object(5)
memory usage: 75.8+ KB
['tweet_id', 'tweet_full_text', 'tweet_entities', 'tweet_entities_hashtags', 'tweet_metadata', 'tweet_retweet_count', 'tweet_favorite_count', 'tweet_favorited', 'tweet_retweeted', 'tweet_lang']


In [481]:
# Explore the data set.
tweets.info()
tweets.describe().T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1174 entries, 0 to 1173
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   tweet_id                 1174 non-null   int64 
 1   tweet_full_text          1174 non-null   object
 2   tweet_entities           1174 non-null   object
 3   tweet_entities_hashtags  1007 non-null   object
 4   tweet_metadata           1174 non-null   object
 5   tweet_retweet_count      1174 non-null   int64 
 6   tweet_favorite_count     1174 non-null   int64 
 7   tweet_favorited          1174 non-null   bool  
 8   tweet_retweeted          1174 non-null   bool  
 9   tweet_lang               1174 non-null   object
dtypes: bool(2), int64(3), object(5)
memory usage: 75.8+ KB


count          mean           std           min  \
tweet_id              1174.0  1.567612e+18  2.427553e+13  1.567574e+18   
tweet_retweet_count   1174.0  8.629472e+00  2.978468e+01  0.000000e+00   
tweet_favorite_count  1174.0  3.713799e-01  2.044700e+00  0.000000e+00   

                               25%           50%           75%           max  
tweet_id              1.567590e+18  1.567611e+18  1.567633e+18  1.567655e+18  
tweet_retweet_count   0.000000e+00  1.000000e+00  3.000000e+00  3.030000e+02  
tweet_favorite_count  0.000000e+00  0.000000e+00  0.000000e+00  4.200000e+01

In [482]:
# Would it be useful to only look at retweeted and favourite tweet messages?
# Explain your answer.
tweets['tweet_retweet_count'].value_counts()
tweets['tweet_favorite_count'].value_counts()

# Retweets and favourite tweets messages are generally a repetition of the original tweet.
# Hence, it is not useful to only look at both of those columns as it makes up the 'noise'
# of the dataset. Based on the value counts, more than half of those counts were empty.
# It will be useful to focus on the original tweet instead. 

0      526
1      215
2      114
3       70
5       35
4       27
7       18
12      16
8       15
73      14
9       13
6       12
208     12
35      10
37       6
11       6
53       5
10       5
63       4
44       4
150      4
85       3
41       3
76       3
62       3
207      3
68       3
24       2
23       2
72       2
78       2
16       2
107      1
79       1
169      1
39       1
57       1
54       1
49       1
48       1
40       1
20       1
19       1
15       1
14       1
13       1
303      1
Name: tweet_retweet_count, dtype: int64

0     1027
1       91
2       16
3       13
4        7
5        5
6        2
13       1
28       1
20       1
18       1
17       1
14       1
10       1
12       1
11       1
9        1
8        1
7        1
42       1
Name: tweet_favorite_count, dtype: int64

In [483]:
# Create a new DataFrame containing only the text.
tweets_text = tweets[['tweet_full_text']]

# View the DataFrame.
tweets_text.head()

tweet_full_text
0  As Arkansas’ first Comprehensive Stroke Certified Center, UAMS provides Arkansans with access to the most advanced stoke care. Join us in our mission to make a difference in the health and well-be...
1                                                              RT @AndreaGrammer: Work-life balance is at the foundation of how decisions are made and where #PremiseHealth is headed.  We're #hiring for…
2                                                           RT @OntarioGreens: $10 billion can go a long way to fixing our broken #Healthcare system.\n\nYet Doug Ford would rather spend it ALL on a hig…
3                                                             RT @modrnhealthcr: 🚨#NEW:🚨 Insurance companies are figuring out the best ways to collect information about members’ race and ethnicity data…
4  ICYMI: Our recent blogs on Cybersecurity in Accounting https://t.co/4nnK0FiVVL and Digital Transformation in Healthcare Finance https://t.co/jIqn52lHD3 are a great read, take a look!\n\n#blogs #di...

In [484]:
# Loop through the messages, and create a list of values containing the # symbol.
tags = []

for y in [x.split(' ') for x in tweets_text['tweet_full_text'].values]:
    for z in y:
        if '#' in z:
            # Change to lowercase.
            tags.append(z.lower())
            
tags = pd.Series(tags)            

In [485]:
# Display the first 30 records.
tags.head(30)

0                                    #healthcare
1                                 #premisehealth
2                                        #hiring
3                                    #healthcare
4                                        🚨#new:🚨
5                                look!\n\n#blogs
6                         #digitaltransformation
7                                 #cybersecurity
8                                    #accounting
9                                       #finance
10                                   #healthcare
11    https://t.co/jrgqeqdme4\n.\n#firstcoastcna
12                                      #cnaexam
13                           #cnaexampreparation
14                                 #jacksonville
15                                  #cnatraining
16                                        #nurse
17                                      #nursing
18                                    #nurselife
19                            #nursepractitioner
20                  

In [486]:
# Convert the series to a DataFrame in preparation for visualisation.
tags_count = pd.DataFrame(tags.value_counts()).reset_index()
# Rename the columns.
tags_count.rename(columns = {'index' : 'word', 
                             0 : 'count'}, inplace=True)

In [487]:
# Fix the count datatype.
tags_count['count'].astype(int)

# View the result.
tags_count.info()

0       716
1        80
2        41
3        40
4        38
       ... 
1749      1
1750      1
1751      1
1752      1
1753      1
Name: count, Length: 1754, dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1754 entries, 0 to 1753
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   word    1754 non-null   object
 1   count   1754 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 27.5+ KB


In [488]:
# Display records where the count is larger than 10.
tags_count.loc[tags_count['count']  > 10]

word  count
0                   #healthcare    716
1                       #health     80
2                     #medicine     41
3                           #ai     40
4                          #job     38
5                      #medical     35
6                     #strategy     30
7               #pharmaceutical     28
8                #digitalhealth     25
9                       #pharma     25
10                   #marketing     25
11                  #medtwitter     24
12     #competitiveintelligence     24
13                     #biotech     24
14                       #meded     23
15                      #hiring     18
16                     #vaccine     18
17                        #news     17
18                  #technology     17
19             #machinelearning     17
20             #womeninmedicine     16
21                       #covid     16
22                 #coronavirus     16
23        #competitivemarketing     16
24                  #healthtech     15
25                    #wellness     15
26  #doctorofveterinarymedicine     14
27                     #covid19     14
28                     #science     14
29                    #medicare     14
30                 #patientcare     13
31                        #tech     13
32                      #foamed     13
33                 #healthcare.     13
34                #telemedicine     13
35                   #education     13
36                      #doctor     12
37                   #hospitals     12
38                #mentalhealth     12
39                         #usa     11
40     drugs\n\n#tipsfornewdocs     11
41                      #nurses     11
42                  #telehealth     11
43                 #healthcare?     11
44       #digitaltransformation     11
45                #healthcareit     11
46                  #innovation     11

In [489]:
# Create a plotly barplot indicating records with a count >10 records.
tags_count_above_10 = tags_count.loc[tags_count['count'] > 10]

In [490]:
# Create the plot.
fig = px.bar(tags_count_above_10,
             x = 'word',
             y = 'count',
             title = 'Top trending hashtags (>10 tweets)',
             text='count')

fig.update_traces(textposition='outside')
fig.update_layout(uniformtext_minsize=10)

# View the barplot.
fig.show();

In [491]:
# Based on the barplot above, noted that there are a few overrepresented hashtags.
# Those hashtags included words like medicine, care and health.
# To remove those hashtags while leaving a main hashtag. (healthcare, health, medicine)
# Filter all hashtags using the keywords above.

tags_count_above_10.loc[tags_count_above_10['word'].str.contains("care", case=False)]
tags_count_above_10.loc[tags_count_above_10['word'].str.contains("medicine", case=False)]
tags_count_above_10.loc[tags_count_above_10['word'].str.contains("health", case=False)]

word  count
0     #healthcare    716
29      #medicare     14
30   #patientcare     13
33   #healthcare.     13
43   #healthcare?     11
45  #healthcareit     11

word  count
2                     #medicine     41
20             #womeninmedicine     16
26  #doctorofveterinarymedicine     14
34                #telemedicine     13

word  count
0      #healthcare    716
1          #health     80
8   #digitalhealth     25
24     #healthtech     15
33    #healthcare.     13
38   #mentalhealth     12
42     #telehealth     11
43    #healthcare?     11
45   #healthcareit     11

In [492]:
# Drop these rows except for the main hashtags.
tags_count_above_10.drop(tags_count_above_10.index[[8,20,24,26,29,30,33,38,34,42,43,45]], inplace=True)

In [493]:
# Rechecking the dataset.
tags_count_above_10

word  count
0                #healthcare    716
1                    #health     80
2                  #medicine     41
3                        #ai     40
4                       #job     38
5                   #medical     35
6                  #strategy     30
7            #pharmaceutical     28
9                    #pharma     25
10                #marketing     25
11               #medtwitter     24
12  #competitiveintelligence     24
13                  #biotech     24
14                    #meded     23
15                   #hiring     18
16                  #vaccine     18
17                     #news     17
18               #technology     17
19          #machinelearning     17
21                    #covid     16
22              #coronavirus     16
23     #competitivemarketing     16
25                 #wellness     15
27                  #covid19     14
28                  #science     14
31                     #tech     13
32                   #foamed     13
35                #education     13
36                   #doctor     12
37                #hospitals     12
39                      #usa     11
40  drugs\n\n#tipsfornewdocs     11
41                   #nurses     11
44    #digitaltransformation     11
46               #innovation     11

In [494]:
# Creating barplot for the remaining most frequently used hashtags.

fig = px.bar(tags_count_above_10,
             x = 'word',
             y = 'count',
             title = 'Top trending hashtags (>10 tweets)',
             text='count')

fig.update_traces(textposition='outside')
fig.update_layout(uniformtext_minsize=10)

# View the barplot.
fig.show();

# 

# Assignment activity 6

### Investigate the main concerns posed by the NHS. 

In [495]:
# Prepare your workstation.

# import datetime.
from datetime import datetime

# Load the appointments_regional.csv file.
ar = pd.read_csv('/Users/shin/Desktop/lse_daca/course_2/LSE_DA201_Assignment_files/appointments_regional.csv') 
# View the DataFrame.
ar.head()

icb_ons_code appointment_month appointment_status hcp_type appointment_mode  \
0    E54000034           2020-01           Attended       GP     Face-to-Face   
1    E54000034           2020-01           Attended       GP     Face-to-Face   
2    E54000034           2020-01           Attended       GP     Face-to-Face   
3    E54000034           2020-01           Attended       GP     Face-to-Face   
4    E54000034           2020-01           Attended       GP     Face-to-Face   

  time_between_book_and_appointment  count_of_appointments  
0                             1 Day                   8107  
1                    15  to 21 Days                   6791  
2                       2 to 7 Days                  20686  
3                    22  to 28 Days                   4268  
4                     8  to 14 Days                  11971

In [496]:
# Print the min and max dates.
# temporarily converting date format to yyyy-mm-dd hh:mm:ss.
ar['appointment_month'] = pd.to_datetime(ar['appointment_month'], format='%Y-%m')
ar['appointment_month'].max()
ar['appointment_month'].min()

Timestamp('2022-06-01 00:00:00')

Timestamp('2020-01-01 00:00:00')

In [497]:
# Filter the data set to only look at data from 2021-08 onwards.
ar_filtered = ar.loc[ar['appointment_month'] >= '2021-08-01 00:00:00']

**Question 1:** Should the NHS start looking at increasing staff levels? 

In [498]:
# Create an aggregated data set to review the different features.
ar_agg = ar_filtered

# View the DataFrame.
ar_agg

icb_ons_code appointment_month appointment_status hcp_type  \
3652      E54000034        2021-08-01           Attended       GP   
3653      E54000034        2021-08-01           Attended       GP   
3654      E54000034        2021-08-01           Attended       GP   
3655      E54000034        2021-08-01           Attended       GP   
3656      E54000034        2021-08-01           Attended       GP   
...             ...               ...                ...      ...   
596816    E54000050        2022-06-01            Unknown  Unknown   
596817    E54000050        2022-06-01            Unknown  Unknown   
596818    E54000050        2022-06-01            Unknown  Unknown   
596819    E54000050        2022-06-01            Unknown  Unknown   
596820    E54000050        2022-06-01            Unknown  Unknown   

       appointment_mode time_between_book_and_appointment  \
3652       Face-to-Face                             1 Day   
3653       Face-to-Face                    15  to 21 Days   
3654       Face-to-Face                       2 to 7 Days   
3655       Face-to-Face                    22  to 28 Days   
3656       Face-to-Face                     8  to 14 Days   
...                 ...                               ...   
596816          Unknown                       2 to 7 Days   
596817          Unknown                    22  to 28 Days   
596818          Unknown                     8  to 14 Days   
596819          Unknown                 More than 28 Days   
596820          Unknown                          Same Day   

        count_of_appointments  
3652                     6553  
3653                     2390  
3654                    10547  
3655                      937  
3656                     4961  
...                       ...  
596816                     21  
596817                      8  
596818                     28  
596819                     17  
596820                     10  

[223418 rows x 7 columns]

In [499]:
# converting appointment_month back to string and correct format, yyyy-mm.
ar_agg['appointment_month'] = ar_agg['appointment_month'].astype(str).str[:7]

# Determine the total number of appointments per month.
ar_df = ar_agg.groupby('appointment_month')[['count_of_appointments']].sum().reset_index()

# Add a new column to indicate the average utilisation of services.
# Monthly aggregate / 30 to get to a daily value.

ar_df['daily_utilisation'] = ar_df['count_of_appointments'].apply(lambda x: ((x / 30) / 1200000) * 100).round(decimals = 1)
# View the DataFrame.
ar_df

appointment_month  count_of_appointments  daily_utilisation
0            2021-08               23852171               66.3
1            2021-09               28522501               79.2
2            2021-10               30303834               84.2
3            2021-11               30405070               84.5
4            2021-12               25140776               69.8
5            2022-01               25635474               71.2
6            2022-02               25355260               70.4
7            2022-03               29595038               82.2
8            2022-04               23913060               66.4
9            2022-05               27495508               76.4
10           2022-06               25828078               71.7

In [500]:
# Plot sum of count of monthly visits.

# Create a lineplot with plotly.
fig = px.line(ar_df, x = 'appointment_month', 
              y = 'count_of_appointments',
              title = 'Number of monthly visits', 
              labels = {'appointment_month': 'Month', 
                        'count_of_appointments': 'Number of appointments'})
fig.show()

In [501]:
# Plot monthly capacity utilisation.


# Create a lineplot.
fig = px.line(ar_df, x = 'appointment_month', 
              y = 'daily_utilisation',
              title = 'Monthly capacity utilisation', 
              labels = {'appointment_month': 'Month',
                        'daily_utilisation':  'Average daily appointments'})
fig.show()

**Question 2:** How do the healthcare professional types differ over time?

In [502]:
# Grouping ar_filtered dataframe by appointment month and hcp_type
# while counting the number of appointments.
ar_hcptype = ar_filtered.groupby(['appointment_month', 'hcp_type'])[['count_of_appointments']].sum().reset_index()

# Creating lineplot
fig = px.line(ar_hcptype, x = 'appointment_month', 
              y = 'count_of_appointments',
              color = 'hcp_type',
              title = 'Total appointment across months and healthcare professional type', 
                 labels = {'appointment_month': 'Month', 
                           'hcp_type': 'Healthcare professional type', 
                           'count_of_appointments': 'Number of appointments'})
fig.show()

**Question 3:** Are there significant changes in whether or not visits are attended?

In [503]:
# Grouping ar_filtered dataframe by appointment month and appointment status
# while counting the number of appointments.
ar_app_status = ar_filtered.groupby(['appointment_month',
                                     'appointment_status'])[['count_of_appointments']].sum().reset_index()


# Create a lineplot.
fig = px.line(ar_app_status, x = 'appointment_month', 
              y = 'count_of_appointments',
              color = 'appointment_status',
              title = 'Total appointment across months and appointment status', 
                 labels = {'appointment_month': 'Month', 
                           'appointment_status': 'Appointment status', 
                           'count_of_appointments': 'Number of appointments'})
fig.show()

**Question 4:** Are there changes in terms of appointment type and the busiest months?

In [504]:
# Grouping ar_filtered dataframe by appointment month and appointment type
# while counting the number of appointments.
ar_app_type = ar_filtered.groupby(['appointment_month',
                                     'appointment_mode'])[['count_of_appointments']].sum().reset_index()


# Create a lineplot.
fig = px.line(ar_app_type, x = 'appointment_month', 
              y = 'count_of_appointments',
              color = 'appointment_mode',
              title = 'Total appointment across months and appointment type', 
                 labels = {'appointment_month': 'Month', 
                           'appointment_mode': 'Appointment mode', 
                           'count_of_appointments': 'Number of appointments'})
fig.show()

**Question 5:** Are there any trends in time between booking an appointment?

In [505]:
# Grouping ar_filtered dataframe by appointment month and time taken for appointment booking
# while counting the number of appointments.
ar_app_time = ar_filtered.groupby(['appointment_month',
                                     'time_between_book_and_appointment'])[['count_of_appointments']].sum().reset_index()


# Create a lineplot.
fig = px.line(ar_app_time, x = 'appointment_month', 
              y = 'count_of_appointments',
              color = 'time_between_book_and_appointment',
              title = 'Total appointment across months and time taken for appointment booking', 
                 labels = {'appointment_month': 'Month', 
                           'time_between_book_and_appointment': 'time taken for appointment booking', 
                           'count_of_appointments': 'Number of appointments'})
fig.show()

**Question 6:** How do the spread of service settings compare?

In [506]:
# Let's go back to the national category DataFrame you created in an earlier assignment activity.
nc = pd.read_excel('/Users/shin/Desktop/lse_daca/course_2/LSE_DA201_Assignment_files/national_categories.xlsx') 

In [507]:
# Create a new DataFrame consisting of the month of appointment and the number of appointments.
nc_app = nc
# View the DataFrame.
nc_app

appointment_date icb_ons_code  \
0            2021-08-02    E54000050   
1            2021-08-02    E54000050   
2            2021-08-02    E54000050   
3            2021-08-02    E54000050   
4            2021-08-02    E54000050   
...                 ...          ...   
817389       2022-06-30    E54000054   
817390       2022-06-30    E54000054   
817391       2022-06-30    E54000054   
817392       2022-06-30    E54000054   
817393       2022-06-30    E54000054   

                             sub_icb_location_name            service_setting  \
0       NHS North East and North Cumbria ICB - 00L       Primary Care Network   
1       NHS North East and North Cumbria ICB - 00L                      Other   
2       NHS North East and North Cumbria ICB - 00L           General Practice   
3       NHS North East and North Cumbria ICB - 00L           General Practice   
4       NHS North East and North Cumbria ICB - 00L           General Practice   
...                                            ...                        ...   
817389              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   
817390              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   
817391              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   
817392              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   
817393              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   

                  context_type                       national_category  \
0       Care Related Encounter  Patient contact during Care Home Round   
1       Care Related Encounter                         Planned Clinics   
2       Care Related Encounter                              Home Visit   
3       Care Related Encounter              General Consultation Acute   
4       Care Related Encounter            Structured Medication Review   
...                        ...                                     ...   
817389  Care Related Encounter             Unplanned Clinical Activity   
817390  Care Related Encounter                         Planned Clinics   
817391  Care Related Encounter              Planned Clinical Procedure   
817392  Care Related Encounter            General Consultation Routine   
817393  Care Related Encounter              General Consultation Acute   

        count_of_appointments appointment_month  
0                           3           2021-08  
1                           7           2021-08  
2                          79           2021-08  
3                         725           2021-08  
4                           2           2021-08  
...                       ...               ...  
817389                     12           2022-06  
817390                      4           2022-06  
817391                     92           2022-06  
817392                      4           2022-06  
817393                     19           2022-06  

[817394 rows x 8 columns]

In [508]:
# Create a boxplot to investigate spread of service settings.
nc_app_ss = nc_app.groupby(['appointment_month',
                            'service_setting'])[['count_of_appointments']].sum().reset_index()

fig = px.box(nc_app_ss,
             x="appointment_month",
             y="count_of_appointments",
             color='service_setting',
             title='Appointments vs Months and Service settings',
             labels = {'appointment_month': 'Month', 
                       'service_setting': 'Service setting', 
                       'count_of_appointments': 'Number of appointments'})
fig.show()

In [509]:
# Create a boxplot to investigate the service settings without GP.

nc_app_ss_1 = nc_app_ss[nc_app_ss['service_setting'].isin(['Extended Access Provision',
                                                           'Other',
                                                           'Primary Care Network', 
                                                           'Unmapped'])]


fig = px.box(nc_app_ss_1,
             x="appointment_month",
             y="count_of_appointments",
             color='service_setting',
             title='Appointments vs Months and Service settings',
             labels = {'appointment_month': 'Month', 
                       'service_setting': 'Service setting', 
                       'count_of_appointments': 'Number of appointments'})

fig.show();

# 

### Provide a summary of your findings and recommendations based on the analysis.

## Findings


While the monthly capacity utilisation spiked to an all time high of 84.5% on Nov 2021, 
healthcare capacity have not reached it's limits.

Average appointment attended ranged between 20 – 30 million.
There were very low number of patients not attending appointments. (< 1.57 million per month)

There is a low number of appointments every Sunday and a high number of appointments every monday for general practice. In addition, the number of appointments generally decreases from Monday to Sunday. This trend for general practice is consistent across all seasons.

This can plausibly be because most patients’ well-being was affected during working (Monday to Friday) and non-working days. (Saturday – Sunday)

For the month of August 2021,

About 50% of national category appointments related to General consultation routine.

Over 90% of context type appointments were related to care related encounters.

Over 90% of service settings appointments were related to general practice.

There were a significant number of patients making appointments on the same day, followed by appointments made within a week, 2 weeks and 1 day in advance.

Majority of patients preferred Face-to-Face and Telephone consultations. NHS can look into reallocating GP appointments to teleconsultations.

The NHS can look into reducing these numbers by investigating further into the the reasons why these appointments are unmapped. In doing so, will allow NHS to be more efficient in allocating resources.

Based on current trending hashtags on twitter, the top 3 hashtags were all related to
Health. (health, healthcare and medicine)



## Recommendations
Reallocate staff on weekends to weekdays

Promote teleconsultation as it reduces wait time, decreases contact health risk and have direct access to different service settings. Tweeting can be a powerful tool to raise awareness about clinical appointments, engage with patients and develop leads. It also increases exposure towards the medical industry. The NHS can plausibly set up a verified business twitter account to raise awareness of the cost of missing medical appointments. In addition, should there be any vital information to alert the people, twitter represents one of the fastest social media platforms to reach the masses.

Leverage on Social media platform to send news to the masses
